File: frontend/lib/srs/sm2.ts
export type Sm2State = {
  interval: number;
  repetitions: number;
  easinessFactor: number;
};

export function calculateNextReview(
  state: Sm2State,
  quality: number,
): Sm2State {
  if (quality < 3) {
    return {
      interval: 1,
      repetitions: 0,
      easinessFactor: state.easinessFactor,
    };
  }

  const repetitions = state.repetitions + 1;
  const interval =
    repetitions === 1 ? 1 : repetitions === 2 ? 6 : Math.round(state.interval * state.easinessFactor);
  const easinessFactor = Math.max(
    1.3,
    state.easinessFactor + (0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)),
  );

  return {
    interval,
    repetitions,
    easinessFactor,
  };
}

File: backend/app/services/sm2.py
def sm2_calculate(rating: int, ease_factor: float, interval: int, repetitions: int):
    """
    Calculate new SM-2 values based on student rating.

    Args:
        rating: 0-5 (0 = forgot, 5 = perfect)
        ease_factor: current EF (1.3-2.5)
        interval: current interval in days
        repetitions: number of successful recalls

    Returns:
        (new_ease_factor, new_interval, new_repetitions)
    """
    if rating < 3:
        # Reset — review tomorrow
        new_interval = 1
        new_repetitions = 0
        new_ease_factor = ease_factor  # unchanged
    else:
        # Calculate new EF
        new_ease_factor = ease_factor + (
            0.1 - (5 - rating) * (0.08 + (5 - rating) * 0.02)
        )
        # Floor guard
        new_ease_factor = max(1.3, new_ease_factor)

        # Calculate new interval
        if repetitions == 0:
            new_interval = 1
        elif repetitions == 1:
            new_interval = 6
        else:
            new_interval = round(interval * new_ease_factor)

        new_repetitions = repetitions + 1

    return new_ease_factor, new_interval, new_repetitions
